release-24.3: client cert expirations cached by usage

angles-n-daemons opened the pull request:
Backport:
  * 1/1 commits from "security: move cert expiry cache into new clientcert package" (#143077)
  * 1/1 commits from "security: change client cert expiration caching eviction policy" (#143384)

Please see individual PRs for details.

/cc @cockroachdb/release

Release justification: required bugfix for exporting client certificate expiration.

angles-n-daemons committed:
security: move cert expiry cache into new clientcert package

To preempt the work for reporting client certificate expiration based on
recently used certificates (#142686), a new `clientcert` package is
being created to house these new changes.

Epic: CRDB-40209

angles-n-daemons committed:
security: change client cert expiration caching eviction policy

Previously, we cached and emitted metrics for client certificates
indefinitely or until they expired. However, this strategy ignores
the case where a team rotates their certificates, discontinuing to
use soon-to-expire ones, in which crdb would continue to report on
their expiration.

This change modifies the caching behavior so that eviction from the
cache is based on last usage. It does a few additional things here:
 - Supports the removal of child metrics by key.
 - No longer uses the  `server.client_cert_expiration_cache.capacity`
         cluster setting.

Fixes: #142686
Epic: CRDB-40209

Release note (ops change): Cluster setting
`server.client_cert_expiration_cache.capacity` no longer used.
Metrics `security.certificate.expiration.client` and
`security.certificate.ttl.client` now report lowest value seen for
user in last 24h.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143591)
<!-- Reviewable:end -->

angles-n-daemons committed:
security: fix possible deadlock in certificate caching code

The changes introduce in #143384 allowed for caching client
certificates based on when they were last seen, expanding
the functionality for reporting on upcoming expirations.

However this introduced a subtle possible deadlock as a
result, for adding a new certificate would lock the cache,
then the metric, while reading the metric (as the metrics
scraper would do) would first lock the metric, then the
cache.

Because these operations could happen independently, and
locked in the inverse order, a scrape which occurred during
an upsert could create a system deadlock.

To fix this, we aim to remove the requirement that the
metrics scrape has on locking the cache.

Fixes: #143662
Epic: CRDB-40209

Release note (bug fix): Fixes possible deadlock in cert
caching code.
